README.md
# Logstash Plugin

[![Travis Build Status](https://travis-ci.org/logstash-plugins/logstash-filter-bytes.svg)](https://travis-ci.org/logstash-plugins/logstash-filter-bytes)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically build documentation for this plugin. We provide a template file, index.asciidoc, where you can add documentation. The contents of this file will be converted into html and then placed with other plugin documentation in a [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting config examples, you can use the asciidoc `[source,json]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
bin/logstash-plugin install --no-verify
```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
bin/logstash-plugin install /your/local/plugin/logstash-filter-awesome.gem
```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.


CHANGELOG.md
## 1.0.3
 - [Doc] Fixes examples table layout [#10](https://github.com/logstash-plugins/logstash-filter-bytes/pull/10)

## 1.0.2
 - Fixes crashing if source is not a String type [#6](https://github.com/logstash-plugins/logstash-filter-bytes/pull/6)

## 1.0.1
  - Remove non-existant configuration option [#5](https://github.com/logstash-plugins/logstash-filter-bytes/pull/5)

## 1.0.0
  - Initial release  


spec/filters/bytes_spec.rb
# encoding: utf-8
require "logstash/devutils/rspec/spec_helper"
require "logstash/filters/bytes"

describe LogStash::Filters::Bytes do
  let(:config) do <<-CONFIG
    filter {
      bytes {
        target => dest
        tag_on_failure => [ "boom", "_bytesparsefailure" ]
      }
    }
  CONFIG
  end

  describe "empty" do
    sample("") do
      expect(subject.get('tags')).to eq(["boom", "_bytesparsefailure"])
    end
  end

  describe "garbage" do
    sample("abcdef") do
      expect(subject.get('tags')).to eq(["boom", "_bytesparsefailure"])
    end
  end

  describe "no number, only units" do
    sample("mb") do
      expect(subject.get('tags')).to eq(["boom", "_bytesparsefailure"])
    end
  end

  describe "no units" do
    sample("0") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(0)
    end

    sample("123") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(123)
    end

    sample("12.3") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(12)
    end
  end

  describe "from bytes" do
    sample("32.8B") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(33)
    end
  end

  describe "from kilobytes" do
    sample("32kB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024)
    end

    sample("32KB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024)
    end
  end

  describe "from megabytes" do
    sample("32mB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024)
    end

    sample("32mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024)
    end

    sample("32MB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024)
    end
  end

  describe "from gigabytes" do
    sample("32gB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024)
    end

    sample("32gb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024)
    end

    sample("32GB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024)
    end
  end

  describe "from terabytes" do
    sample("32tB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024)
    end

    sample("32tb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024)
    end

    sample("32TB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024)
    end
  end

  describe "from petabytes" do
    sample("32pB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024 * 1024)
    end

    sample("32pb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024 * 1024)
    end

    sample("32PB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024 * 1024)
    end
  end

  describe "from exabytes" do
    sample("32eB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024)
    end

    sample("32eb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024)
    end

    sample("32EB") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024)
    end
  end

  describe "with spaces" do
    sample("32 mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024)
    end

    sample("32\tmb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1024 * 1024)
    end
  end

  describe "using metric conversion system" do
    let(:config) do <<-CONFIG
      filter {
        bytes {
          target => dest
          conversion_method => metric
        }
      }
    CONFIG
    end

    sample("32 mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(32 * 1000 * 1000)
    end
  end

  describe "no decimal separator" do
    sample("3 mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(3 * 1024 * 1024)
    end      

    sample("3,124 mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(3124 * 1024 * 1024)
    end      
  end

  describe "digits only to right of decimal separator" do
    sample(".3124 mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq((0.3124 * 1024 * 1024).round)
    end
  end

  describe "digits only to left of decimal separator" do
    sample("3. mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(3 * 1024 * 1024)
    end      

    sample("3,124. mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq(3124 * 1024 * 1024)
    end      
  end

  describe "digits on both sides of decimal separator" do
    sample("3.56 mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq((3.56 * 1024 * 1024).round)
    end      

    sample("3,124.56 mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq((3124.56 * 1024 * 1024).round)
    end      
  end

  describe "non-default decimal separator" do
    let(:config) do <<-CONFIG
      filter {
        bytes {
          target => dest
          decimal_separator => ','
        }
      }
    CONFIG
    end

    sample("3,9 mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq((3.9 * 1024 * 1024).round)
    end

    sample("1.9 kb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq((19 * 1024).round)
    end

    sample("3.124,56 mb") do
      expect(subject).to include("dest")
      expect(subject.get('dest')).to eq((3124.56 * 1024 * 1024).round)
    end
  end

  describe "with two decimal separators" do
    sample("1.000.000 mb") do
      expect(subject.get('tags')).to eq(["boom", "_bytesparsefailure"])
    end

    context "non-default decimal separator" do
      let(:config) do <<-CONFIG
        filter {
          bytes {
            target => dest
            decimal_separator => ','
            tag_on_failure => [ "boom", "_bytesparsefailure" ]
          }
        }
      CONFIG
      end

      sample("1,000,000 mb") do
        expect(subject.get('tags')).to eq(["boom", "_bytesparsefailure"])
      end
    end
  end
end


lib/logstash/filters/bytes.rb
# encoding: utf-8
require "logstash/filters/base"
require "logstash/namespace"

# This filter will parse a given string as a computer storage
# value (e.g. "123 MB" or "5.3GB") and add a new numeric field with
# size in bytes
#
class LogStash::Filters::Bytes < LogStash::Filters::Base

  PREFIX_POWERS = {
    'k' => 1, # 1 kilobyte = 1024 ^ 1 bytes
    'm' => 2, # 1 megabyte = 1024 ^ 2 bytes
    'g' => 3, # 1 gigabyte = 1024 ^ 3 bytes
    't' => 4, # 1 terabyte = 1024 ^ 4 bytes
    'p' => 5, # 1 petabyte = 1024 ^ 5 bytes
    'e' => 6, # 1 exabyte = 1024 ^ 6 bytes
  }.freeze

  DIGIT_GROUP_SEPARATORS = " _,."

  # Setting the config_name here is required. This is how you
  # configure this filter from your Logstash config.
  #
  # filter {
  #   bytes {
  #     source => "123 MB"
  #     target => "num_bytes"
  #   }
  # }
  #
  config_name "bytes"
  
  # Source string, e.g. "123 MB", "5.3GB", etc.
  config :source, :validate => :string, :default => "message"

  # Target field name
  config :target, :validate => :string

  # Conversion method, either "binary" (1K = 1024B) or "metric" (1K = 1000B)
  config :conversion_method, :validate => [ "binary", "metric" ], :default => "binary"

  # Decimal separator
  config :decimal_separator, :validate => [ ".", "," ] , :default => "."

  # Append values to the `tags` field when there has been no
  # successful match
  config :tag_on_failure, :validate => :array, :default => ["_bytesparsefailure"]

  private
  def normalize_number(number)
    number
      .tr("^0-9#{@decimal_separator}", '')
      .tr(@decimal_separator, '.')
  end

  public
  def register
    # Add instance variables 
  end # def register

  public
  def filter(event)

    source = event.get(@source)

    if !source or !source.is_a? String
      @tag_on_failure.each{|tag| event.tag(tag)}
      return
    end
    source.strip!

    # Parse the source into the number part (e.g. 123),
    # the unit prefix part (e.g. M), and the unit suffix part (e.g. B)
    match = source.match(/^([0-9#{DIGIT_GROUP_SEPARATORS}#{@decimal_separator}]*)\s*([kKmMgGtTpPeE]?)([bB]?)$/)
    if !match
      @tag_on_failure.each{|tag| event.tag(tag)}
      return
    end

    number, prefix, suffix = match.captures

    # Flag error if more than one decimal separator is found
    num_decimals = number.count(@decimal_separator)
    if num_decimals > 1
      @tag_on_failure.each{|tag| event.tag(tag)}
      return
    end

    number = normalize_number(number.strip)
    if number == ''
      @tag_on_failure.each{|tag| event.tag(tag)}
      return
    end

    if suffix == ''
      suffix = 'B'
    end

    # Convert the number to bytes
    result = number.to_f
    if prefix != ''
      if @conversion_method == 'binary'
        result *= (1024 ** PREFIX_POWERS[prefix.downcase])
      else
        result *= (1000 ** PREFIX_POWERS[prefix.downcase])
      end
    end
    result = result.round

    event.set(@target, result)

    # filter_matched should go in the last line of our successful code
    filter_matched(event)
  end # def filter
end # class LogStash::Filters::Bytes
